Comorphisms/CASL2Prenex.hs

```diff
@@ -0,0 +1,150 @@
+preprocessSen :: CASLFORMULA -> CASLFORMULA
+preprocessSen sen = case sen of
+ Relation sen1 Equivalence sen2 _ -> let -- remove equivalences
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Con [Junction Dis [Negation sen1' nullRange, sen2'] nullRange,
+                   Junction Dis [Negation sen2' nullRange, sen1'] nullRange
+                  ] nullRange
+ Relation sen1 Implication sen2 _ -> let -- remove implications
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Dis [Negation sen1' nullRange, sen2'] nullRange
+ Relation sen1 RevImpl sen2 _ -> let -- remove reverse implications
+   sen1' = preprocessSen sen1
+   sen2' = preprocessSen sen2
+  in Junction Dis [Negation sen2' nullRange, sen1'] nullRange
+ Junction j sens _ -> let -- only binary junctions
+   sens' = map preprocessSen sens
+   bindLeft [] = Junction j [] nullRange
+   bindLeft [x] = Junction j [x] nullRange
+   bindLeft [x,y] = Junction j [x,y] nullRange
+   bindLeft (x:y:s') = Junction j [x, bindLeft (y:s')] nullRange
+  in bindLeft sens'
+ Quantification q vdecls qsen _ -> let -- recursion
+   qsen' = preprocessSen qsen
+  in Quantification q vdecls qsen' nullRange
+ _ -> sen
+
+-- function 2: computePNF
+-- see page 292 in LPL.pdf
+-- but instead of checking that a variable occurs free in a formula
+-- I do a alpha renaming such that this property is ensured
+
+computePNF :: Int -> CASLFORMULA -> (Int, CASLFORMULA)
+computePNF n sen = case sen of
+ Negation nsen _ -> let
+   (n', nsen') = computePNF n nsen
+   (n'', nsen'') = case nsen' of
+    Quantification q vdecls qsen _ -> -- deMorgan
+       (n', Quantification (dualQuant q) vdecls
+              (Negation qsen nullRange) nullRange)
+    _ -> (n', Negation nsen' nullRange)
+  in (n'', nsen'')
+ Junction j [sen1, sen2] _ -> let
+   (n', sen1') = computePNF n sen1
+   (n'', sen2') = computePNF n' sen2
+   (n3, sen') = case (sen1', sen2') of
+     (Quantification q vdecls qsen _, aSen) -> let
+         vars = flatVAR_DECLs vdecls
+         (n''', vars') = getFreshVars vars n''
+         vdecls' = map (\(v,s) -> mkVarDecl v s) $ map fst vars'
+         qsen' = foldl (\f (v,s,t)-> substitute v s t f) qsen $
+                        map (\((x,y),(z,t)) -> (z, t, Qual_var x y nullRange))
+                        vars'
+         (n4, qsen'') = computePNF n''' (Junction j [qsen', aSen] nullRange)
+        in (n4, Quantification q vdecls' qsen'' nullRange)
+     (aSen, Quantification q vdecls qsen _) -> let
+         vars = flatVAR_DECLs vdecls
+         (n''', vars') = getFreshVars vars n''
+         vdecls' = map (\(v,s) -> mkVarDecl v s) $ map fst vars'
+         qsen' = foldl (\f (v,s,t)-> substitute v s t f) qsen $
+                        map (\((x,y),(z,t)) -> (z, t, Qual_var x y nullRange))
+                        vars'
+         (n4, qsen'') = computePNF n''' (Junction j [aSen, qsen'] nullRange)
+        in (n4, Quantification q vdecls' qsen'' nullRange)
```

Comment:
these two cases share a lot of code. This can be avoided by treating n-ary junctions in general. Look for the first quantification in the list of con/disjuncts and process it.